index.html

```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
       <p>The underlying source is supposed to be kept alive between the transfer and transfer-receiving steps, or as long as the data holder is alive.
       In a sense, between these steps, the data holder is attached to the underlying source as if it was a track.</p>
     </div>
+    <h3>MediaStreamTrack frame counters</h3>
+    <div>
+      <p>
+        Frame counters allow you to tell what the capture frame rate is, which
+        may be different from what is expected from settings - for instance, the
+        camera may be slowing down frame production because it's dark or the
+        system may be CPU starved and therefore dropping frames.
+      </p>
+    </div>
+    <div>
+      <pre class="idl"
+>partial interface MediaStreamTrack {
+  Promise&lt;MediaStreamTrackStats&gt; getStats();
+};
+</pre>
+    </div>
+    <div>
+      <p>At creation of a {{MediaStreamTrack}} object, <var>track</var>, run the
+      following steps:</p>
+      <ol>
+        <li>
+          <p>Initialize <var>track</var>.`[[FramesCaptured]]` to 0.</p>
+        </li>
+      </ol>
+    </div>
+    <div>
+      <p>Monitor frames that arrive to <var>track</var> from its source by
+      running the following in-parallel steps for every <var>frame</var>:</p>
+      <ol>
+        <li>
+          <p>If <var>frame</var> is to be discarded by <var>track</var> in
+          order to achive the frame rate decimation of <var>track</var>'s
+          settings, abort these steps.</p>
```

Comment:
I hesitate introducing algorithms here, since this seems deep in implementation territory.

Algorithms are nice when dealing with JS-observable operations, but it's not JS-observable when frames "arrive to" to a MediaStreamTrack, whatever that means.

This might be better expressed as requirements.

Again, linking to the frameRate constraint may be helpful (if this feature is tied to it existing).

As [noted](https://w3c.github.io/mediacapture-main/#media-track-supported-constraints) in the spec: _"The constraints specified in this specification apply only to instances of [MediaStreamTrack](https://w3c.github.io/mediacapture-main/#dom-mediastreamtrack) generated by [MediaDevices](https://w3c.github.io/mediacapture-main/#dom-mediadevices).[getUserMedia](https://w3c.github.io/mediacapture-main/#dom-mediadevices-getusermedia)(), unless stated otherwise in other specifications."_

Reply:
> I hesitate introducing algorithms here, ...

> This might be better expressed as requirements.

I tried without an algorithm but there as confusion, then Youenn suggested an internal slot to make things less confusing. Considering the amount of back and forth, I don't think this is a bad idea. But the important thing is everyone is on the same page with what we're doing.

> ...it's not JS-observable when frames "arrive to" a MediaStreamTrack, whatever that means.

It's not a JS-observable event, which should give implementations some amount of wiggle-room. I'm just trying to count frames at a sensible time. How would this be phrased otherwise?

Note that there is a difference between counting in the capture processes or renderer process etc, hence the "arrive at the track" language, even though that is not the final destination.

Reply:
I changed the PR to be requirements-based